### PR TITLE
Keep original file mode when overwriting

### PR DESCRIPTION
**Pull request by a-sk**
Hi, a stumbled across a bug, when after I call `atomic_write` with overwrite=True, I got file with a different mode than before.

### CODE REVIEW

File: atomicwrites/__init__.py [hunk 1]
```diff
@@ -20,6 +20,8 @@ def _path_to_unicode(x):
 
 if sys.platform != 'win32':
     def _replace_atomic(src, dst):
+        if os.path.exists(dst):
+            os.chmod(src, os.stat(dst).st_mode)
```
Comment (untitaker) [inline atomicwrites/__init__.py:24]: Could you store the mode somewhere, and then use `chmod` after the rename, so we don't have to check whether the file exists?

File: tests/test_atomicwrites.py [hunk 2]
```diff
@@ -40,6 +41,17 @@ def test_replace_simultaneously_created_file(tmpdir):
     assert len(tmpdir.listdir()) == 1
 
 
+def test_replace_keep_mode_unchanged(tmpdir):
+    fname = tmpdir.join('ha')
+    fname.write('abc')
+    os.chmod(str(fname), 0640)
```
Comment (untitaker) [inline tests/test_atomicwrites.py]: This is invalid syntax in Python 3, use `0o640`.

File: atomicwrites/__init__.py (hunk 1 above)
Comment (a-sk) [inline atomicwrites/__init__.py:24]: Actually, I did this deliberately.
I called `os.chmod` in my code, where I use atomicwrite, but some applications,
look at file `mtime` (in my case it is apache httpd rewrite map) and cache the result, but `os.chmod` does not change the mtime, so I had to update `mtime` as well.

File: tests/test_atomicwrites.py (hunk 2 above)
Comment (a-sk) [inline tests/test_atomicwrites.py]: Oh, sorry, I pushed the update

File: atomicwrites/__init__.py (hunk 1 above)
Comment (untitaker) [inline atomicwrites/__init__.py:24]: I just noticed that in the case of `_replace_atomic` the target file already exists.

**Comment by untitaker** [comment]
This change makes `_replace_atomic` non-atomic, which defeats the entire purpose of this package. If there's some flag we can pass to `MoveFileEx` that preserves mode, we'd be golden.

**Comment by a-sk** [comment]
Maybe I don't fully understand the term `atomic` here, or did you mean that this change makes `_replace_atomic` not consistent across platforms?

> If there's some flag we can pass to MoveFileEx that preserves mode, we'd be golden.

I don't have windows machine to test, but you mean, that by default `MoveFileExW` does not keep the source access rights?

**Comment by untitaker** [comment]
@a-sk I think it would be problematic if somebody issues a chmod on `dst` (manually, outside of python-atomicwrites), and some other program uses python-atomicwrites at the same time:
- python-atomicwrites gets `stat` of `dst`.
- Somebody changes permissions of `dst`.
- python-atomicwrites restores the _old_ permissions on `src`.
- python-atomicwrites calls `rename`, restores the _old_ permissions on `dst`.

You have similar problems even if you restore the permissions after you `rename`d. So restoring permissions and `renaming` should happen in one [atomic](https://en.wikipedia.org/wiki/Atomicity_%28database_systems%29) step.

**Comment by untitaker** [comment]
> but you mean, that by default MoveFileExW does not keep the source access rights?

I have no idea, I didn't think about file permissions at all when writing this code.

**Comment by untitaker** [comment]
> or did you mean that this change makes _replace_atomic not consistent across platforms?

Oh, I didn't even consider that one. Good point...

**Comment by a-sk** [comment]
> @a-sk I think it would be problematic if somebody issues a chmod on dst (manually, outside of python-atomicwrites), and some other program uses python-atomicwrites at the same time:

oh, now I see the problem here :(

**Comment by untitaker** [comment]
I'm going to close this because neither on POSIX nor on Windows you can do `rename` and `chmod` in one step. In some situations this doesn't matter, but in those you can also do the `chmod` yourself.

**Comment by a-sk** [comment]
Yeah, I came to the same conclusion, sorry for bothering.

**Comment by untitaker** [comment]
No problem, I think you raised an important point regarding file permissions. I've added a note to the README.

**Pull request closed without merging**